File: src/main/kotlin/com/iver/blpslab1/config/CustomLoginModule.java
package com.iver.blpslab1.config;

import javax.security.auth.Subject;
import javax.security.auth.callback.CallbackHandler;
import javax.security.auth.login.LoginException;
import javax.security.auth.spi.LoginModule;
import java.security.Principal;
import java.util.Map;

public class CustomLoginModule implements LoginModule {

    private Subject subject;
    private CallbackHandler callbackHandler;
    private Map<String, ?> sharedState;
    private Map<String, ?> options;

    private Principal userPrincipal;

    @Override
    public void initialize(Subject subject, CallbackHandler callbackHandler, Map<String, ?> sharedState, Map<String, ?> options) {
        this.subject = subject;
        this.callbackHandler = callbackHandler;
        this.sharedState = sharedState;
        this.options = options;
    }

    @Override
    public boolean login() throws LoginException {
//        if (callbackHandler == null) {
//            throw new LoginException("No callback handler is available");
//        }
//
//        NameCallback nameCallback = new NameCallback("Username:");
//        PasswordCallback passwordCallback = new PasswordCallback("Password:", false);
//
//        try {
//            callbackHandler.handle(new Callback[] { nameCallback, passwordCallback });
//        } catch (UnsupportedCallbackException ex) {
//            throw new LoginException(ex.getMessage());
//        } catch (Exception ex) {
//            throw new LoginException(ex.getMessage());
//        }
//
//        String username = nameCallback.getName();
//        char[] password = passwordCallback.getPassword();
//
//        // perform authentication here and set userPrincipal if authenticated
//        boolean authenticated = authenticateUser(username, password);
//        if (authenticated) {
//            userPrincipal = new CustomPrincipal(username);
//        } else {
//            throw new LoginException("Authentication failed");
//        }

        return true;
    }

    private boolean authenticateUser(String username, char[] password) {
        // perform authentication logic here
        // return true if authenticated, false otherwise
        return true;
    }

    @Override
    public boolean commit() throws LoginException {
        if (userPrincipal == null) {
            return false;
        } else {
            subject.getPrincipals().add(userPrincipal);
            return true;
        }
    }

    @Override
    public boolean abort() throws LoginException {
        return true;
    }

    @Override
    public boolean logout() throws LoginException {
        subject.getPrincipals().remove(userPrincipal);
        return true;
    }

}

File: src/main/kotlin/com/iver/blpslab1/config/CustomLoginModule.kt
package com.iver.blpslab1.config

import com.sun.security.auth.UserPrincipal
import org.springframework.security.authentication.BadCredentialsException
import org.springframework.security.core.userdetails.UserDetailsService
import org.springframework.security.crypto.bcrypt.BCryptPasswordEncoder
import org.springframework.security.crypto.password.PasswordEncoder
import java.io.IOException
import javax.security.auth.Subject
import javax.security.auth.callback.*
import javax.security.auth.login.LoginException
import javax.security.auth.spi.LoginModule


class CustomLoginModule(private val userDetailsService: UserDetailsService) : LoginModule {
    private var callbackHandler: CallbackHandler? = null
    private var subject: Subject? = null
    private var userPrincipal: UserPrincipal? = null
    override fun initialize(
        subject: Subject, callbackHandler: CallbackHandler, sharedState: Map<String?, *>?,
        options: Map<String?, *>?
    ) {
        this.subject = subject
        this.callbackHandler = callbackHandler
    }

    @Throws(LoginException::class)
    override fun login(): Boolean {
        val nameCallback = NameCallback("Username:")
        val passwordCallback = PasswordCallback("Password:", false)
        val callbacks: Array<Callback> = arrayOf(nameCallback, passwordCallback)
        try {
            callbackHandler!!.handle(callbacks)
        } catch (e: IOException) {
            throw LoginException("Error in callback: " + e.message)
        } catch (e: UnsupportedCallbackException) {
            throw LoginException("Error in callback: " + e.message)
        }
        val username = nameCallback.name
        val password: String = String(passwordCallback.password)
        val userDetails = userDetailsService.loadUserByUsername(username)
        if (passwordEncoder().matches(password, userDetails.password)) {
            userPrincipal = UserPrincipal(username)
            return true
        }
        throw BadCredentialsException("Invalid username/password")
    }

    @Throws(LoginException::class)
    override fun commit(): Boolean {
        subject!!.principals.add(userPrincipal)
        return true
    }

    @Throws(LoginException::class)
    override fun abort(): Boolean {
        return true
    }

    @Throws(LoginException::class)
    override fun logout(): Boolean {
        subject!!.principals.remove(userPrincipal)
        return true
    }

    private fun passwordEncoder(): PasswordEncoder {
        return BCryptPasswordEncoder()
    }
}